components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,574 @@
+///      - Initializing the schema for a new database
+///      - Upgrading the schema for an existing database
+///      - Extra preparation/finishing steps, for example setting up SQLite functions
+///
+///    - Call open_database() in your database constructor:
+///      - If the database file is not present, open_database() will create a new DB and call prepare(),
+///        init(), then finish()
+///      - If the database file exists, open_database() will open it and call prepare(),
+///        upgrade_from() for each upgrade that needs to be applied, then finish().
+///       - If there is an error opening/migrating the database and
+///        ErrorHandling::DeleteAndRecreate was specified, open_database() will delete the
+///        database file and try to start from scratch
+///
+///  See the autofill DB code for an example.
+///
+use crate::ConnExt;
+use rusqlite::{Connection, OpenFlags, NO_PARAMS};
+use std::path::PathBuf;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    // Generic error meaning that something went wrong during the migration.  You can return this
+    // from your upgrade functions to signal that the database is beyond repair and can't be
+    // migrated.
+    #[error("MigrationError: {0}")]
+    MigrationError(String),
+    // Error with the migration logic struct, for example the number upgrade functions doesn't
+    // line up with start_version and end_version
+    #[error("MigrationLogicError: {0}")]
+    MigrationLogicError(String),
+    #[error("Database version too old: {0}")]
+    VersionTooOld(u32),
+    #[error("Database version too new: {0}")]
+    VersionTooNew(u32),
+    #[error("Error executing SQL: {0}")]
+    SqlError(#[from] rusqlite::Error),
+    #[error("IOError")]
+    IOError(#[from] std::io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub enum DatabaseLocation {
+    File(PathBuf),
+    Memory,
+}
+
+impl DatabaseLocation {
+    fn exists(&self) -> bool {
+        match self {
+            DatabaseLocation::Memory => false,
+            DatabaseLocation::File(path) => path.exists(),
+        }
+    }
+
+    fn open(&self, open_flags: OpenFlags) -> Result<Connection> {
+        match self {
+            DatabaseLocation::Memory => Ok(Connection::open_in_memory_with_flags(open_flags)?),
+            DatabaseLocation::File(path) => Ok(Connection::open_with_flags(path, open_flags)?),
+        }
+    }
+
+    fn delete(&self) -> Result<()> {
+        match self {
+            DatabaseLocation::Memory => (),
+            DatabaseLocation::File(path) => {
+                if path.exists() {
+                    std::fs::remove_file(path)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub enum ErrorHandling {
+    // Try to start over with a fresh database.  This is the behavior that most components should
+    // use (See SYNC-937, SYNC-816)
```

Comment:
minor nit, but IMO we shouldn't link to a jira that's only accessible to staff. IOW, please link to the public github issue - just #xxxx is OK.